accounts-db/src/tiered_storage.rs

```diff
@@ -109,29 +115,32 @@ impl TieredStorage {
         skip: usize,
         format: &TieredStorageFormat,
     ) -> TieredStorageResult<Vec<StoredAccountInfo>> {
-        if self.is_read_only() {
-            return Err(TieredStorageError::AttemptToUpdateReadOnly(
-                self.path.to_path_buf(),
-            ));
-        }
-
-        if format == &HOT_FORMAT {
-            let result = {
-                let writer = HotStorageWriter::new(&self.path)?;
-                writer.write_accounts(accounts, skip)
-            };
-
-            // panic here if self.reader.get() is not None as self.reader can only be
-            // None since we have passed `is_read_only()` check previously, indicating
-            // self.reader is not yet set.
-            self.reader
-                .set(TieredStorageReader::new_from_path(&self.path)?)
-                .unwrap();
+        let was_written = self.already_written.swap(true, Ordering::AcqRel);
+
+        // If it was not previously written, and the current thread has
+        // successfully updated the already_written flag, then the current
+        // thread will proceed and writes the input accounts.
+        if !was_written {
+            if format == &HOT_FORMAT {
+                let result = {
+                    let writer = HotStorageWriter::new(&self.path)?;
+                    writer.write_accounts(accounts, skip)
+                };
+
+                // panic here if self.reader.get() is not None as self.reader can only be
+                // None since we have passed `is_read_only()` check previously, indicating
+                // self.reader is not yet set.
+                self.reader
+                    .set(TieredStorageReader::new_from_path(&self.path)?)
+                    .unwrap();
+
+                return result;
+            }
 
-            return result;
+            Err(TieredStorageError::UnknownFormat(self.path.to_path_buf()))
+        } else {
+            panic!("TieredStorage::write_accounts() is expected to be invoked once per file.");
```

Comment:
Good suggestion.  Will do the second one as it's easier to add new formats.

Reply:
So here's what I've got.   I am doing the handle-all-branches approach as we will later add COLD_FORMAT.  But since CI does not allow me to do an `else` right after an `if` (it suggests doing `else if` instead, which falls a bit like the mixed approach).  Let me know if it looks cleaner to you or not.

```
         let was_written = self.already_written.swap(true, Ordering::AcqRel);

         if was_written {
             panic!("cannot write same tiered storage file more than once");
         }

         if format == &HOT_FORMAT {
             let result = {
                 let writer = HotStorageWriter::new(&self.path)?;
                 writer.write_accounts(accounts, skip)
             };

             // panic here if self.reader.get() is not None as self.reader can only be
             // None since a false-value `was_written` indicates the accounts file has
             // not been written previously, implying is_read_only() was also false.
             debug_assert!(!self.is_read_only());
             self.reader
                 .set(TieredStorageReader::new_from_path(&self.path)?)
                 .unwrap();

             result
         } else {
             Err(TieredStorageError::UnknownFormat(self.path.to_path_buf()))
         }
```